src/hsm_cl.rs

```diff
@@ -0,0 +1,134 @@
+use class_group::primitives::cl_dl_lcm::{CLDLProofPublicSetup, Witness, HSMCL, PK};
+use curv::arithmetic::traits::Converter;
+use curv::arithmetic::traits::Samplable;
+use curv::elliptic::curves::traits::ECPoint;
+use curv::elliptic::curves::traits::ECScalar;
+use curv::BigInt;
+use curv::FE;
+use curv::GE;
+
+pub use class_group::primitives::cl_dl_lcm::Ciphertext;
+pub type PublicKey = PK;
+
+pub type Proof = CLDLProofPublicSetup;
+
+lazy_static::lazy_static! {
+     pub static ref PUBLIC_SETUP: BigInt = BigInt::from(b"A2LPoC".as_ref());
+}
+
+#[derive(Clone, Debug)]
+pub struct KeyPair(HSMCL);
+
+impl KeyPair {
+    pub fn public_key(&self) -> &PublicKey {
+        &self.0.pk
+    }
+
+    pub fn gen() -> Self {
+        Self(HSMCL::keygen_with_setup(&FE::q(), &1348, &*PUBLIC_SETUP))
+    }
+}
+
+pub fn encrypt(public_key: &PublicKey, message: &crate::secp256k1::KeyPair) -> (Ciphertext, Proof) {
+    let r = BigInt::sample_below(&(&public_key.stilde * BigInt::from(2).pow(40)));
+    let x = BigInt::from(message.secret_key().serialize().as_ref());
+    let ciphertext = HSMCL::encrypt_predefined_randomness(&public_key, &x, &r);
+    let pk_untagged_bytes = &message.public_key().serialize()[1..];
+    let X = GE::from_bytes(pk_untagged_bytes).unwrap();
+
+    let proof = CLDLProofPublicSetup::prove(
+        Witness { x: &x, r },
+        &public_key,
+        &ciphertext,
+        &X,
+        &*PUBLIC_SETUP,
```

Comment:
As described above, this implicit dependency can be avoided by keeping the state in the public key.